### ISSUE TITLE
Introduce `get_absolute_url` for SearchResult instances

### ISSUE THREAD

**Issue by kezabelle** [django-haystack/django-haystack#967]
Mostly, to output a link for a search result, it is necessary to either store the `get_absolute_url` of the `Model` instance into the `SearchIndex`, usually via the `model_attr` or whatever, or call `{{ result.object.get_absolute_url }}` in the template, which in the naive case of not using `load_all()`, will trigger a database query per `SearchResult` iterated (the template usage is what occurs in the docs).

From a cursory glance, SearchResult could hit the 80% use case in the typical 80/20 rule, given the following stipulations:
- if the `SearchResult` instance's `_object` is not None, it has been populated, probably via a `load_all()` call, so it would be a database no-op to call `get_absolute_url` on `_object`
  - I think. Would need verification.
- if the `_object` is None;
  - get the `model` property,
    - if we can't get a model, fail early.
    - if the model does not have a `get_absolute_url`, fail early.
  - get the model's local, non-virtual field names via `_meta.get_fields_with_model` (probably filtering out relation fields, and calling `get_attname` to find out the kwarg)
  - for every key in `SearchResult.get_additional_fields`, if it matches one of the attr names given by `get_fields_with_model`, keep it.
  - instantiate a new `model` using the intersection of the `get_additional_fields` and `get_fields_with_model` (ie: those we kept)
  - try and call get_absolute_url with the unsaved model instance, catching `NoReverseMatch` and failing with the same sentinel value as any fail-early condition (eg: `None`, `''`)

This would allow for many common URL naming patterns (using a slug or pk as the unique identifier in the URI) to avoid needing to load anything from the database, either because it already has been (`load_all`), or because the `SearchResult` was able to instantiate enough of an object to re-create the necessary args/kwargs for urlconf  reversal, without the user/template designer needing to know about the SQS's invocation, or implementation.

Instantiation of partial objects when calling the `SearchResult.get_absolute_url` multiple times for a single `SearchResult` could be minimized by either using `cached_property`, or storing it into the `SearchResult` on successful reverse.

**Comment by westurner** [comment]
:+1:

**Comment by westurner** [comment]
``load_all`` docs:
- http://django-haystack.readthedocs.io/en/v2.5.0/searchqueryset_api.html#SearchQuerySet.load_all
- http://django-haystack.readthedocs.io/en/v2.5.0/best_practices.html#avoid-hitting-the-database

**Comment by SteveByerly** [comment]
This sounds like a really inefficient way to go about it. Object instantiation is not always cheap, especially if there are a lot of objects.

Perhaps there is an alternative approach that defines the get_absolute_url on the index and exposes the method to the result.

**Comment by westurner** [comment]
Minimally, ``SearchResult.get_absolute_url`` could proxy to ``Model.get_absolute_url`` if ``SearchResult`` contains the necessary parameters for ``get_absolute_url``.

**Comment by westurner** [comment]
e.g. with ``Model.get_absolute_url(search_result)``

**Comment by SteveByerly** [comment]
yeah, i'm down for anything that doesn't require instantiating an object.
